tests/sklearn/test_sklearn_autolog.py

```diff
@@ -0,0 +1,366 @@
+import functools
+import inspect
+from mock import mock
+
+import numpy as np
+import pytest
+import sklearn
+
+import mlflow.sklearn
+from mlflow.utils.mlflow_tags import MLFLOW_PARENT_RUN_ID
+from mlflow.sklearn.utils import _get_arg_names
+
+FIT_FUNC_NAMES = ["fit", "fit_transform", "fit_predict"]
+TRAINING_SCORE = "training_score"
+ESTIMATOR_CLASS = "estimator_class"
+ESTIMATOR_NAME = "estimator_name"
+
+
+def get_iris():
+    iris = sklearn.datasets.load_iris()
+    return iris.data[:, :2], iris.target
+
+
+def fit_model(model, Xy, fit_func_name):
+    if fit_func_name == "fit":
+        model.fit(*Xy)
+
+    if fit_func_name == "fit_transform":
+        model.fit_transform(*Xy)
+
+    if fit_func_name == "fit_predict":
+        model.fit_predict(*Xy)
+
+    return model
+
+
+def get_run(run_id):
+    return mlflow.tracking.MlflowClient().get_run(run_id)
+
+
+def get_run_data(run_id):
+    client = mlflow.tracking.MlflowClient()
+    data = client.get_run(run_id).data
+    # Ignore tags mlflow logs by default (e.g. "mlflow.user")
+    tags = {k: v for k, v in data.tags.items() if not k.startswith("mlflow.")}
+    artifacts = [f.path for f in client.list_artifacts(run_id)]
+    return data.params, data.metrics, tags, artifacts
+
+
+def load_model_by_run_id(run_id):
+    return mlflow.sklearn.load_model("runs:/{}/model".format(run_id))
+
+
+def stringify_dict_values(d):
+    return {k: str(v) for k, v in d.items()}
+
+
+@pytest.fixture(params=FIT_FUNC_NAMES)
+def fit_func_name(request):
+    return request.param
+
+
+def test_autolog_preserves_original_function_attributes():
+    def get_func_attrs(f):
+        attrs = {}
+        for attr_name in ["__doc__", "__name__"]:
+            if hasattr(f, attr_name):
+                attrs[attr_name] = getattr(f, attr_name)
+
+        attrs["__signature__"] = inspect.signature(f)
+        return attrs
+
+    def get_cls_attrs(cls):
+        attrs = {}
+        for method_name in FIT_FUNC_NAMES:
+            if hasattr(cls, method_name):
+                attr = getattr(cls, method_name)
+                if isinstance(attr, property):
+                    continue
+
+                attrs[method_name] = get_func_attrs(attr)
+        return attrs
+
+    before = [get_cls_attrs(cls) for _, cls in mlflow.sklearn.utils._all_estimators()]
+    mlflow.sklearn.autolog()
+    after = [get_cls_attrs(cls) for _, cls in mlflow.sklearn.utils._all_estimators()]
+
+    for b, a in zip(before, after):
+        assert b == a
+
+
+@pytest.mark.skipif(
+    not mlflow.sklearn.utils._is_old_version(), reason="This test fails on sklearn>=0.20.3"
+)
+def test_autolog_emits_warning_on_older_versions_of_sklearn():
+    with pytest.warns(
+        UserWarning, match="Autologging utilities may not work properly on scikit-learn"
+    ):
+        mlflow.sklearn.autolog()
+
+
+def test_autolog_does_not_terminate_active_run():
+    mlflow.sklearn.autolog()
+    mlflow.start_run()
+    sklearn.cluster.KMeans().fit(*get_iris())
```

Comment:
After calling `sklearn.utils.all_estimators`, we can sklearn's submodules without importing them.